add a phase transition to the inner core material model

jdannberg opened the pull request:
This is the first one of (hopefully) a series of pull requests that will extend the inner core cookbook, adding a number of new features to it. Once we have all of the new functionality, we will also modify the cookbook text and add an input file that demonstrates how to use these new features. 

This pull request adds the option to include a phase transition with variable depth/width/Clapyeron slope/etc. in the model, similarly to the latent heat material model, just much simpler.

jdannberg commented:
I just noticed this is not going to work in the innermost region of the inner core model (where the gravity and hence the pressure gradient is zero), so I have to think about a fix for this.

tjhei commented on cookbooks/inner_core_convection/inner_core_convection.cc:
```diff
@@ -60,6 +79,32 @@ namespace aspect
+      double phase_func;
+      // use delta function for width = 0
+      if (transition_width == 0)
+        (depth_deviation > 0) ? phase_func = 1 : phase_func = 0;
```
this is missing an assignment ``phase_func = ``

bangerth replied:
Yes, I would agree -- legal, but hard to read :-)

jdannberg replied:
Oh, right. I just copied that from another material model. I'll make a separate pull request to fix it there too.

bangerth commented on cookbooks/inner_core_convection/inner_core_convection.cc:
```diff
@@ -111,6 +157,39 @@ namespace aspect
           }
           prm.leave_subsection();
+
+          prm.declare_entry ("Phase transition depth", "0.0",
```
do the inner core people state locations in terms of depth from the surface, or in terms of distance from the center?

jdannberg replied:
I asked, and it looks like the modellers state locations in terms of distance from the centre, and the seismologists in terms of distance from the inner core boundary. I changed everything to use the radius instead of depth, which is useful anyway, as eventually we want to let the core grow over time...

jdannberg commented:
Okay, I addressed your comments and fixed the part about the pressure gradient at the centre of the model. As this will only be used in a cookbook anyway, I would prefer to only add a test for the final new cookbook setup that includes all of the new code once we've finished that.

bangerth commented on cookbooks/inner_core_convection/inner_core_convection.cc:
```diff
@@ -60,6 +103,64 @@ namespace aspect
+      // We need to convert the depth to pressure,
+      // keeping in mind that for this model, the pressure is only the dynamic pressure,
+      // so we have to get the hydrostatic pressure explicitly as an input.
+      const double radius = this->get_geometry_model().maximal_depth() - this->get_geometry_model().depth(position);
```
Can you add an assertion that the radius is `>=0`?

bangerth commented on cookbooks/inner_core_convection/inner_core_convection.cc:
```diff
@@ -60,6 +103,64 @@ namespace aspect
+      const double pressure_deviation = hydrostatic_pressure(radius) - hydrostatic_pressure(transition_radius);
+
+      double depth_deviation = transition_radius - radius;
+      if (depth_deviation != 0.0)
```
In floating point arithmetic, this condition is always true. You will want to compare against something that is small.

jdannberg commented:
I think I addressed all comments.

bangerth commented on cookbooks/inner_core_convection/inner_core_convection.cc:
```diff
@@ -60,6 +103,68 @@ namespace aspect
+      const double pressure_deviation = hydrostatic_pressure(radius) - hydrostatic_pressure(transition_radius);
+
+      double depth_deviation = transition_radius - radius;
+      if (std::fabs(pressure_deviation) > 100.0 * std::numeric_limits<double>::epsilon())
```
I think I should have explained better. If your pressure is, say, 100GPa, i.e., 1e11, then the smallest difference you can compute in line 127 is 1e-16*1e11=1e-5. This is *always* bigger than 100*eps.

To make this `if` statement meaningful, you need to make the tolerance *relative* to the size of quantities you are comparing. So maybe something like
```
if (std::fabs(pressure_deviation) 
   > 
   100.0 * std::numeric_limits<double>::epsilon() * 
   (hydrostatic_pressure(radius) + hydrostatic_pressure(transition_radius))/2)
```
i.e., 100*eps times the *average* of the two pressures.

jdannberg replied:
I see your point, and I've changed the line accordingly. 
However, keep in mind that this model is nondimensionalized, and that the pressure is only the dynamic pressure. So the pressures are all very small, and might as well become negative, and the only thing I want to make sure is that we do not divide by zero in the next line.

bangerth replied:
Ah, I hadn't realized that it's non-dimensional. Either way, even in non-dimensional models, quantities may be on the order of 1e-6 or 1e6, and you could run into the same problem.

But the new version looks good.
